File: frontend/src/types/user_availability.ts
import type { UserSimple } from './user'; // Assuming UserSimple exists for embedding

/**
 * Base interface for UserAvailability data (common fields).
 */
export interface UserAvailabilityBase {
  available_from: string; // ISO 8601 datetime string
  available_to: string;   // ISO 8601 datetime string
  note?: string | null;
}

/**
 * Interface for creating or updating UserAvailability.
 */
export interface UserAvailabilityCreateUpdate extends UserAvailabilityBase {}

/**
 * Interface representing a UserAvailability record received from the API.
 */
export interface UserAvailability extends UserAvailabilityBase {
  id: number;
  user_id: number;
  slot_id: number;
  user: UserSimple; // Embed simplified user info
  created_at: string; // ISO 8601 datetime string
  updated_at: string; // ISO 8601 datetime string
} 

File: backend/src/app/schemas/user_availability.py
from pydantic import BaseModel, ConfigDict, field_validator
from datetime import datetime
from typing import Optional

# Import UserSimple for embedding user info
from .user import UserSimple 

# Base schema with common attributes
class UserAvailabilityBase(BaseModel):
    available_from: datetime
    available_to: datetime
    note: Optional[str] = None

    @field_validator('available_to')
    def check_dates(cls, v, values):
        # Pydantic v2: 'values' might be passed differently, adjust if needed
        # This depends on exact Pydantic version and how validation context is handled
        # For simplicity, let's assume 'available_from' is already validated and present
        # A more robust approach might involve model-level validation
        if 'available_from' in values.data and v <= values.data['available_from']:
            raise ValueError('available_to must be after available_from')
        return v

# Schema for creating/updating user availability (input)
# user_id will come from the current authenticated user
# slot_id will come from the path parameter
class UserAvailabilityCreateUpdate(UserAvailabilityBase):
    pass

# Schema for reading user availability (output)
class UserAvailability(UserAvailabilityBase):
    id: int
    user_id: int
    slot_id: int
    user: UserSimple # Embed simplified user information
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True) 